src/sage/rings/function_field/drinfeld_modules/finite_drinfeld_module.py

```diff
@@ -231,51 +259,232 @@ def frobenius_charpoly(self, var='X'):
 
         ::
 
-            sage: trace = phi.frobenius_trace()
-            sage: trace
-            (4*z3^2 + 6*z3 + 3)*T + 3*z3^2 + z3 + 4
-            sage: norm = phi.frobenius_norm()
-            sage: norm
-            (5*z3^2 + 2*z3)*T^2 + (4*z3^2 + 3*z3)*T + 5*z3^2 + 2*z3
+            sage: phi.frobenius_charpoly(algorithm="NotImplemented")
+            Traceback (most recent call last):
+            NotImplementedError: Algorithm "NotImplemented" not implemented
+
+        ALGORITHM:
+
+        By default, this method uses the so-called *crystalline*
+        algorithm which computes the characteristic polynomial of the
+        Frobenius acting on the crystalline cohomology of the Drinfeld
+        module. For further details, see [Ang1997]_. Currently, the only
+        alternative is to use the *gekeler* approach based on solving
+        the linear system given by `t^{nr} + \sum_{i=0}^{r-1}
+        \phi_{A_{i}}t^{ni} = 0`. For more details, see [Gek2008]_.
+        """
+        methods = [method for method in dir(self)
+                   if method.startswith('_frobenius_charpoly_')]
+        method_name = f'_frobenius_charpoly_{algorithm}'
+        if method_name in methods:
+            if self._frobenius_charpoly is not None:
+                return self._frobenius_charpoly
+            self._frobenius_charpoly = getattr(self, method_name)(var)
+            return self._frobenius_charpoly
+        raise NotImplementedError(f'Algorithm \"{algorithm}\" not implemented')
```

Comment:
```suggestion
        raise NotImplementedError(f'algorithm \"{algorithm}\" is not implemented')
```
Error messages should stat with a noncapitalize letter.